pandas/core/groupby/generic.py

```diff
@@ -1011,6 +1011,10 @@ def _wrap_applied_output(
             result = self.obj._constructor(
                 index=self.grouper.result_index, columns=data.columns
             )
+            for i in range(len(data.index.names)):
```

Comment:
Hmm why doesn't the line above not return the correct result and index levels still need to be added?

Reply:
@mroeschke 
because `grouper.result_index` doesn't have the original indexes it only contains the result index from the method, so I had to append the original indexes to the result.

Reply:
I think `grouper.result_index` should have the correct index without modification after so I think it's better to investigate why `result_index` doesn't return the correct index here.

Reply:
@mroeschke 
I cannot edit `grouper.result_index` because it's used in many other places and it forms the index by the grouper only, which is the usual behavior of `groupby` but when using `resample` or `dropna` it's different as we want the index to contain both the old indexes and the grouper. 
if the data frame isn't empty it will use `reshape.concat` and add both `result_index` and the current indexes, but we cannot use  `reshape.concat` on an empty data frame that's why we use the wrap function. However, the wrap function isn't 100% mimicking the behavior of the concat now it needs this additional for loop to add the current indexes.

Reply:
This was my take. I don't think it's feasible to edit the result from grouper.result_index because it's designed to be like that and the wrap_function is designed to fix the behavior when applying the method on an empty dataframe.

Reply:
`result_index` is perhaps poorly named - it is only the the index for aggregations with `as_index=True`. This is why it's giving the wrong result for transformations.

Reply:
After the line `result = self.obj._constructor(columns=data.columns)`, I think we can call `result = self._set_result_index_ordered(result)`. This is what various transforms use to set the index.

Reply:
I missed what function this code was on; we only want to do this when the function is a transform coming from pandas internals. I think we can add `is_transform` to `_wrap_applied_output` and use that to determine whether or not to call `_set_result_index_ordered`.

Reply:
@rhshadrach using `_wrap_applied_output` is dropping the `groupby` keys.

Reply:
I see - I agree that `_wrap_applied_output` is returning a different shape when the input is empty or non-empty in the datetime test. However, this wrapping code is also used when the user supplies an arbitrary UDF to groupby.apply. We can't make any assumptions about what that UDF does, so I do not think we should be making this modification here.

Can you fix the case of an empty input in `DatetimeIndexResampler._wrap_result`? This is called after the groupby portion is used.

Reply:
@rhshadrach 

This is what I did firstly but @mroeschke said that we should edit the root error by seeing why `groupby.apply` doesn't return the right index. Here is the discussion link: https://github.com/pandas-dev/pandas/pull/47672/commits/e683177e6adb7566bbe88204d70983d84f2504b5

Reply:
@mroeschke - what do you think of my assessment in https://github.com/pandas-dev/pandas/pull/47672#discussion_r936085977 (cf https://github.com/pandas-dev/pandas/pull/47672#discussion_r924772766)

Reply:
Thanks for your assessment @rhshadrach. I think I had gotten mislead earlier by existing naming when you mentioned `result_index` is only used for aggregations when `result_index=True`

Sure it makes sense now that the fix should be applied to `_wrap_result`